File: src/main/kotlin/wowsft/model/gameparams/ship/Ship.java
package wowsft.model.gameparams.ship;

import com.fasterxml.jackson.annotation.*;
import wowsft.config.WoWSFT;
import wowsft.model.gameparams.TypeInfo;
import wowsft.model.gameparams.ship.abilities.AbilitySlot;
import wowsft.model.gameparams.ship.component.ShipComponent;
import wowsft.model.gameparams.ship.component.airdefense.Aura;
import wowsft.model.gameparams.ship.component.artillery.Turret;
import wowsft.model.gameparams.ship.component.torpedo.Launcher;
import wowsft.model.gameparams.ship.upgrades.ShipUpgradeInfo;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

@WoWSFT
@JsonIgnoreProperties(ignoreUnknown = true)
public class Ship
{
    private LinkedHashMap<String, Object> tempComponents = new LinkedHashMap<>();
    private ShipComponent components = new ShipComponent();

    @JsonAlias("ShipAbilities")
    private LinkedHashMap<String, AbilitySlot> shipAbilities;
    @JsonAlias("ShipUpgradeInfo")
    private ShipUpgradeInfo shipUpgradeInfo;

    private float apDamageLimitCoeff;
    private BattleLevels battleLevels;
    private boolean canEquipCamouflage;
    private String defaultCrew;
    private String group;
    private long id;
    private String index;
    @JsonAlias("isPaperShip")
    private boolean paperShip;
    private int level;
    private int maxEquippedFlags;
    private String name;
    private String navalFlag;
    private boolean needShowProjectYear;
    private String peculiarity;
    private List<Float> steerAngle;
    private TypeInfo typeinfo;
    private int weight;

    private String realShipType;
    private String realShipTypeId;
    private String fullName;
    private boolean research;
    private String prevShipIndex;
    private String prevShipName;
    private int prevShipXP;
    private int prevShipCompXP;
    private String typeImage;
    private String imageSmall;

    private LinkedHashMap<String, String> planes = new LinkedHashMap<>();

    private List<List<Consumable>> consumables = new ArrayList<>();
    private List<List<Modernization>> upgrades = new ArrayList<>();
    private int upgradesRow;
    @JsonIgnore
    private List<Integer> selectConsumables = new ArrayList<>();
    @JsonIgnore
    private List<Integer> selectUpgrades = new ArrayList<>();
    @JsonIgnore
    private List<Integer> selectSkills = new ArrayList<>();
    @JsonIgnore
    private int selectSkillPts;
    private LinkedHashMap<String, String> modules = new LinkedHashMap<>();
    private LinkedHashMap<String, Integer> positions = new LinkedHashMap<>();

    private Commander commander;

    private List<Turret> turrets;
    private List<Launcher> launchers;

    private List<Aura> auraFar = new ArrayList<>();
    private List<Aura> auraFarBubble = new ArrayList<>();
    private List<Aura> auraMedium = new ArrayList<>();
    private List<Aura> auraNear = new ArrayList<>();

    private float adrenaline;
    private boolean arUse;

    @JsonSetter
    public void setRealShipType(String realShipType)
    {
        if (StringUtils.isNotEmpty(realShipType)) {
            this.realShipType = realShipType;
            this.realShipTypeId = realShipType.toUpperCase();
            if ("Premium".equalsIgnoreCase(realShipType)) {
                this.realShipTypeId = "FILTER_PREMIUM";
            }
        }
    }

    @JsonAnySetter
    public void setUpTempComponents(String name, Object value)
    {
        tempComponents.put(name, value);
    }

    public String getTypeImage()
    {
        if (typeinfo != null && StringUtils.isNotEmpty(typeinfo.getSpecies()) && StringUtils.isNotEmpty(realShipType)) {
            return "https://cdn.wowsft.com/images/vehicles/types/" + typeinfo.getSpecies() + (realShipType.equalsIgnoreCase("Premium") ? "/premium" : "/normal") + ".png";
        }
        return "";
    }

    public String getImageSmall()
    {
        if (StringUtils.isNotEmpty(index)) {
            return "https://cdn.wowsft.com/images/vehicles/ship_previews/" + index + ".png";
        }
        return "";
    }
}

File: src/main/kotlin/wowsft/model/gameparams/ship/Ship.kt
package wowsft.model.gameparams.ship

import com.fasterxml.jackson.annotation.*
import wowsft.config.WoWSFT
import wowsft.model.gameparams.TypeInfo
import wowsft.model.gameparams.commander.Commander
import wowsft.model.gameparams.consumable.Consumable
import wowsft.model.gameparams.modernization.Modernization
import wowsft.model.gameparams.ship.abilities.AbilitySlot
import wowsft.model.gameparams.ship.component.ShipComponent
import wowsft.model.gameparams.ship.component.airdefense.Aura
import wowsft.model.gameparams.ship.component.artillery.Turret
import wowsft.model.gameparams.ship.component.torpedo.Launcher
import wowsft.model.gameparams.ship.upgrades.ShipUpgradeInfo

import java.util.ArrayList
import java.util.LinkedHashMap

@WoWSFT
@JsonIgnoreProperties(ignoreUnknown = true)
class Ship
{
    var tempComponents = LinkedHashMap<String, Any>()
    var components = ShipComponent()

    @JsonAlias("ShipAbilities")
    var shipAbilities = LinkedHashMap<String, AbilitySlot>()
    @JsonAlias("ShipUpgradeInfo")
    var shipUpgradeInfo = ShipUpgradeInfo()

    var apDamageLimitCoeff = 0.0
    var battleLevels = BattleLevels()
    var canEquipCamouflage = false
    var defaultCrew: String? = null
    var group: String? = null
    var id: Long = 0
    var index: String? = null
    @JsonAlias("isPaperShip")
    var paperShip = false
    var level = 0
    var maxEquippedFlags = 0
    var name: String? = null
    var navalFlag: String? = null
    var needShowProjectYear = false
    var peculiarity: String? = null
    var steerAngle = ArrayList<Double>()
    var typeinfo : TypeInfo? = null
    var weight = 0

    var realShipType: String? = null
        set(realShipType) {
            if (!realShipType.isNullOrBlank()) {
                field = realShipType
                realShipTypeId = if ("Premium".equals(realShipType, ignoreCase = true)) "FILTER_PREMIUM" else realShipType.toUpperCase()
            }
        }
    var realShipTypeId: String? = null
    var fullName: String? = null
    var research = false
    var prevShipIndex: String? = null
    var prevShipName: String? = null
    var prevShipXP = 0
    var prevShipCompXP = 0
    val typeImage: String
        get() = if (!typeinfo?.nation.isNullOrBlank() && !typeinfo?.species.isNullOrBlank() && !realShipType.isNullOrBlank()) {
            "https://cdn.wowsft.com/images/vehicles/types/" + typeinfo?.species + (if (realShipType.equals("Premium", ignoreCase = true)) "/premium" else "/normal") + ".png"
        } else ""
    val imageSmall: String
        get() = if (!index.isNullOrBlank()) "https://cdn.wowsft.com/images/vehicles/ship_previews/$index.png" else ""
    var planes = LinkedHashMap<String, String>()

    var consumables = ArrayList<MutableList<Consumable>>()
    var upgrades = ArrayList<MutableList<Modernization>>()
    var upgradesRow = 0
    @JsonIgnore
    var selectConsumables = ArrayList<Int>()
    @JsonIgnore
    var selectUpgrades = ArrayList<Int>()
    @JsonIgnore
    var selectSkills = ArrayList<Int>()
    @JsonIgnore
    var selectSkillPts = 0
    var modules = LinkedHashMap<String, String>()
    var positions = LinkedHashMap<String, Int>()

    var commander = Commander()

    var turrets = ArrayList<Turret>()
    var launchers = ArrayList<Launcher>()

    var auraFar = ArrayList<Aura>()
    var auraFarBubble = ArrayList<Aura>()
    var auraMedium = ArrayList<Aura>()
    var auraNear = ArrayList<Aura>()

    var adrenaline = 0.0
    var arUse = false
    var steerDist = 0.0

    @JsonAnySetter
    fun setUpTempComponents(name: String, value: Any) {
        tempComponents[name] = value
    }
}
